doc: update the sample template (#2632)
nrfconnect/sdk-nrf: doc/nrf/templates/sample_README.rst

.. _sample:

Template: Sample
################

.. note::
   * Give the sample a concise name that corresponds to the folder name.

>      If the sample targets a specific protocol or chip, add it in the title before the sample name (for example, "NFC:" or "nRF9160:").

Reviewer: ```suggestion
     If the sample targets a specific protocol or device, add it in the title before the sample name (for example, "NFC:" or "nRF9160:").
```

     Do not include the word "sample" in the title.
   * Put the documentation inside the sample folder and use the file name :file:`README.rst`.
   * Use the provided stock phrases and includes when possible.
   * Sections with * are optional and can be left out.
     All other sections are required for all samples.
     Do not add new sections (unless in the sections that allow for further subsections) without discussion with the tech writer team.

The Sample sample demonstrates how to blink LEDs in the rhythm of the music that is played.

.. tip::
   Explain what this sample demonstrates in one, max two sentences (full sentences, not sentence fragments).
   This introduction should give users a clear idea of what the sample can be used for.

   Think about use cases:

>    "The XYZ sample shows how to use the XYZ library" is not a good introduction, because customers do not want to use the XYZ library, but they want to get a task done.

Reviewer: ❤️

   * Mention additional requirements after the table.
   * If SPM is included in the sample, add ``.. include:: /includes/spm.txt`` to include the standard text that states this.

The sample supports the following development kits:

.. include:: /includes/boardname_tables/sample_boardnames.txt
   :start-after: set1_start
   :end-before: set1_end

The sample also requires ...

.. include:: /includes/spm.txt

Overview
********

You can use this sample as a starting point to implement a disco light.

The sample uses the ``:ref:RST link`` library to control the LEDs.
In addition, it uses the ``:ref:RST link`` sound sensor and hooks up to some Internet service to download cool blinking sequences.

.. tip::
   Continue the explanation on what this sample is about.
   What does it show, and why should users try it out?
   What is the practical use?
   How can users extend this sample?
   What libraries are used (link to them)?

Some title*
===========
.. note::
   Add subsections for technical details.
   Give them a suitable title (sentence style capitalization, thus only the first word capitalized).
   If there is nothing important to point out, do not include any subsections.

> The sample application repeatedly calls function ABC, which ...

Reviewer: ```suggestion
The sample repeatedly calls function ABC, which ...
```

.. tip::

>    Do not state what people can see and understand from looking at the code.

Reviewer: This is a bit of a risky statement. Developers who are familiar with their code will always assume that people can understand more from just looking at the code than they actually can. I think putting it like this might lead a dev to not describe something that would clarify how the sample works because they will assume it's obvious.

Author: Clarified.

> User interface*

Reviewer 2: Can we also include CLI commands and UART commands in this section? Not exactly hardware pieces of interface, but they serve a similar purpose.

See the following samples for examples:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/openthread/ncp/README.html
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/openthread/cli/README.html

***************
.. note::
   Add button and LED assignments here, plus other information related to the user interface (for example, supported commands).

Button 1:
   Does something.

Button 2:
   Toggles something.

LED 1:
   On when connected.

LED 2:
   Indicates something.

Configuration*
**************

.. note::
   Always include this section if your sample can be configured by the user.
   Start with the stock phrase that is included with ``|config|``.

|config|

Setup*
======

.. note::
   Add information about the initial setup here, for example, that the user must install or enable some library before they can compile this sample, or set up and select a specific backend.
   Most samples do not need this section.

Configuration options*
======================

.. note::
   * You do not need to list all configuration options of the sample, but only the ones that are important for the sample and make sense for the user to know about.

>    * The syntax below allows to link to sample configuration options in the same way as library configuration options (``:option:`SAMPLE_CONFIG```, which results in :option:`SAMPLE_CONFIG`).

Reviewer 3: ```suggestion
   * The syntax below allows sample configuration options to link to the option descriptions in the same way as the library configuration options link to the corresponding Kconfig descriptions (``:option:`SAMPLE_CONFIG```, which results in :option:`SAMPLE_CONFIG`).
```

User interface*
***************
.. note::

>    Add button and LED assignments here, plus other information related to the user interface.

Reviewer 2: Thread CLI example mentions here a link to CLI command list. Maybe the note should reflect this and allow for some other options than buttons and LEDs?

Author: Added.

Button 1:
   Does something.

Button 2:
   Toggles something.

LED 1:
   On when connected.

LED 2:
   Indicates something.

Configuration*
**************

.. note::
   Always include this section if your sample can be configured by the user.
   Start with the stock phrase that is included with ``|config|``.

|config|

> Setup*

Reviewer 2: I'm always confused about the Setup section, and I see some developers confused too. In nRF5 SDK, it would contain the path to source files, but this path could also be mentioned in Configuration. So do we really need to specify this Setup section?

Maybe Configuration could describe the required configuration steps (or none), with Configuration options describing the non-mandatory ones? Because I feel we have too many similar section here.

Author: We haven't had a setup section in NCS before, so I don't see a reason for anybody to get confused about this, especially since it clearly says that most samples won't need the section. In NCS, the path to the sample is mentioned in the Building and running section, because it isn't a part of the configuration.

The idea is to have a place for required setup steps that must be done before you can test the sample. This is NOT about setting configuration options (which is covered in the following sections), but an initial setup - for example, installing an external library (like here: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/nrf9160/gps/README.html#using-the-sample-with-the-supl-client-library ) or setting up a server or account (like here: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/nrf9160/http_application_update/README.html#setting-up-an-aws-s3-bucket ).

Tried to clarify the note.

.. option:: SAMPLE_CONFIG - Sample configuration

   The sample configuration defines ...

.. option:: ANOTHER_CONFIG - Another configuration

   This configuration option specifies ...

Additional configuration*
=========================

.. note::
   * Add this section to describe and link to any library configuration options that might be important to run this sample.
     You can link to options with ``:option:`CONFIG_XXX```.
   * You do not need to list all possible configuration options, but only the ones that are relevant.

Check and configure the following library options that are used by the sample:

* :option:`CONFIG_BT_DEVICE_NAME` - Defines the Bluetooth device name.
* :option:`CONFIG_BT_GATT_LBS` - Enables the :ref:`lbs_readme`.

Configuration files*
====================

.. note::
   Add this section if the sample provides predefined configuration files.

The sample provides predefined configuration files for typical use cases.
You can find the configuration files in the :file:`XXX` directory.

The following files are available:

* :file:`filename.conf` - Specific scenario.
  This configuration file ...

> Building and running

Reviewer 2: Thread and Zigbee samples also include a line about enabling the stack before testing (+ link to stack configuration page).

Author: But the samples are configured to include the stack by default, or aren't they? If they are NOT configured correctly by default, this information should go into the Setup section.
If they are configured correctly by default, I think that line should be moved to after the build_and_run.txt and be reworded to something like "Note that the sample enables the XXX stack. See \<link to UG>." I added that line.

If the LEDs do not start blinking, check if the music is loud enough.

Known issues and limitations*
*****************************

The sample only works with good music.

> References*

Reviewer 2: I'm a bit concerned that these references might have more sense within overview or some other section before. Should we then duplicate the reference here?

Author: No, I don't think we should duplicate. But we have a few samples that have additional references (mostly to specs).
